### ISSUE TITLE
Firefox can’t establish a connection to the server at ws://localhost:9000/sockjs/173/unq1r1jj/websocket.

### ISSUE THREAD

**Issue by MartiniHenry1988** [meteor/meteor#9255]
I am using Angular 1.3 + Meteor 1.5.2.
On my local machine in browser console I am getting this error:

`Firefox can’t establish a connection to the server at ws://localhost:9000/sockjs/173/unq1r1jj/websocket.       ddp-client.js:1528:18`

My web application is working fine except a bit slow page load times.

But why is this error occurring? Is this something I should worry about? Is it decreasing my web app performance or any other issue which I am not aware of?

**Comment by hwillson** [comment]
@MartiniHenry1988 A few questions:

- What version of Firefox are you using?
- Is that message repeated or do you just see it once?
- Does that message show up again when you refresh the page?
- Is this something that just started happening?
- What do you see when you try to access that URL directory, but use HTTP instead; so try accessing http://localhost:9000/sockjs/173/unq1r1jj/websocket from your browser - do you see `Not a valid websocket request`?
- Are you able to provide a small, runnable reproduction showing this happening?

**Comment by hwillson** [comment]
We haven't heard back regarding the questions asked in https://github.com/meteor/meteor/issues/9255#issuecomment-338679144, so closing for now. Thanks!

**Comment by Mandy94** [comment]
Hi @hwillson ,
I have the same problem and since I havn't found lots of posts about it I would love to open this issue.
Answers to your questions in my case are:
* mozzila Firefox 62.0
* No, only once on lodaing and that's it
* ^
* Not sure if i understood question.
* I see "unable to connect"
* Not sure If i understood but this is what's showing in browser.

![github-orchestator-issue](https://user-images.githubusercontent.com/20498749/45753492-5caf5980-bc19-11e8-8d27-e10dc7c8c918.png)

**Comment by HelmZalee** [comment]
Wondering what fraction of occurrence is reported here.  one/year probably isn't worth worrying about, but here goes anyway.  btw, I think this is a problem with NPM and parcel-bundler, not ff, though I am pretty lost in all this stuff.

FF version: _70.0.1 (32-bit)_
Error: _Firefox can’t establish a connection to the server at ws://localhost:54390/. hmr-runtime.js:29:11
10:19:53.406_
Repeats: __The message is repeated only if the page is refreshed. It does not occur generally, but only with this page._
Line and context: …
_var parent = module.bundle.parent;
if ((!parent || !parent.isParcelRequire) && typeof WebSocket !== 'undefined') {
  var hostname = process.env.HMR_HOSTNAME || location.hostname;
  var protocol = location.protocol === 'https:' ? 'wss' : 'ws';
  var ws = new WebSocket(protocol + '://' + hostname + ':' + process.env.HMR_PORT + '/'); **<== Line 29**
  ws.onmessage = function(event) {
    checkedAssets = {};
    assetsToAccept = [];_
The real problem I'm trying to solve is that FF cannot find the onload handler referenced in the HTML.  That handler is in the javascript file marked "main" in package.json, and the problem is not manifested until after the error described here.  I have no idea whether they are related or not.

**Comment by softwarecreations** [comment]
I've been having this issue constantly, for YEARS with many versions of firefox.
It happens if I open localhost:3000 in my browser.

And it also happens, even if I setup a Nginx SSL reverse proxy with a real, valid letsencrypt certificate and open `https://my.host.com` in the browser, and the meteor server is started with
`export ROOT_URL='https://my.host.com'`
It happens on my local server and on my dedicated server with public IP and letsencrypt SSL etc.

How to reproduce:
Open meteor app and if you don't get the error in the console, just click the address bar and hit enter, if the error doesn't appear, do it again and usually within 3-5 loads the error will appear.

```
Firefox can’t establish a connection to the server at wss://my.host.com/sockjs/378/w3xsrfna/websocket. sockjs-0.3.4.js:1275:14
The connection to wss://my.host.com/sockjs/378/w3xsrfna/websocket was interrupted while the page was loading. sockjs-0.3.4.js:1275:14
```
Current version of firefox is Firefox Developer edition 78.0b1 (64-bit), but the specific version of firefox is completely irrelevant, it's been happening for years on various editions of Firefox.

for Nginx I'm using this
```
map $http_upgrade $connection_upgrade { # WebSocket support
 default upgrade;
 '' '';
}
server {
  ssl_certificate /foo/letsencrypt-certs/fullchain.pem;
  ssl_certificate_key /foo/letsencrypt-certs/privkey.pem;
  include conf.d/ssl_cipher_stuff;

  server_name my.host.com;
  root /foo/bar;

	index index.php index.html;

  location / {
    proxy_pass http://internal-server.addr:3000/;

    #whether I set the http version or not does not affect the error
    # proxy_http_version 1.1;

    proxy_set_header Host $server_name:$server_port;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade; # WebSocket support
    proxy_set_header Connection $connection_upgrade; # WebSocket support

    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_buffering off;
    proxy_connect_timeout 600;
    proxy_send_timeout    600;
    proxy_read_timeout    86400;
    send_timeout          600;
    add_header Content-Language en;
  }

  location ^~ /.well-known/ { }
}
```

**Comment by softwarecreations** [comment]
@hwillson @denihs @sebakerckhof @filipenevola @hexsprite can we please reopen this issue?

**Comment by sebakerckhof** [comment]
What catches my attention is 'just click the address bar and hit enter, if the error doesn't appear, do it again and usually within 3-5 loads this error will appear'
This seems a bit like a race condition like this old FF bug: https://bugzilla.mozilla.org/show_bug.cgi?id=712329

I'll try to see if there's a workaround short-term. Might be a good time as well to upgrade the sockjs library.

It seems like @benjamn started this: https://github.com/meteor/sockjs-client but it didn't make it in yet. So not sure what the state of that is, but I could pick it up.

**Comment by softwarecreations** [comment]
@sebakerckhof Thanks for reopening this issue!
Well spotted regarding that Firefox bug. I've just done some testing that confirms that the issue we're having is described in that bug.

**I just did these tests**

1. Close firefox completely, load meteor app, close firefox. **Repeated 10X with no errors.**
2. Click address bar, hit enter. Error appeared on the 1st reload.

**This issue has nothing to do with self-signed certificates or SSL**
By the way when searching for this error I've come across discussion about wss connections with self-signed certificates. So I just want to clarify that this issue is definitely not at all related to self signed certificates.

Meteor's websocket server does not use SSL (Meteor does not use wss:// it's ws://) so there are no self-signed certs involved with Meteor itself.
But you might be mislead if you connect to localhost:3000 because firefox will show all localhost connections in the network tab with a padlock icon. But that just means the connection is "secure" because it's local... it's not using wss:// if you look at the websocket URL for localhost:3000 it's actually ws://

Also interesting, related to the "secure" padlock icon. Let's say your computer's LAN IP is 192.168.0.123. If you start meteor like this: `meteor --port 0.0.0.0:3000 run` you'll be able to open meteor webapp locally with `http://192.168.0.123:3000` and then Firefox won't show the "secure" padlock icon. (and of course the websocket url will be ws://)

The bottom line is if the tab reloads, you've got a 2~3/5 chance of getting the error. If you open a new tab or open the browser fresh, you definitely won't get the error.

My testing through SSL has all been with a legit valid LetsEncrypt certificate. If anyone uses self signed they might get issues where firefox won't like websockets through that, but that's not related to this bug.

**Comment by sebakerckhof** [comment]
Yes, it has nothing to do with SSL. In the meantime I tested with the latest sockjs version, but it doesn't fix it.

**Comment by softwarecreations** [comment]
What do you think the solution would be? Testing a different websocket server library?

**Comment by MartiniHenry1988** [comment]
I am getting this even now with Meteor 10.x. I stopped worrying about this since it wasn't effecting my application that much (Although I still face longer page loading times on web client). 

But I wonder if it really is effecting performance or causing some connection issues.
Because one thing for fact, connection to the server can be frustratingly unstable in real-world use specially when using Android DDP for Native apps. 

For some reason MDG still has no official DDP client libraries for Native Web, IOS,Android Applications

**Comment by softwarecreations** [comment]
This is still an issue with
```
METEOR@1.11.1
ddp@1.4.0
ddp-client@2.3.3
ddp-common@1.4.0
ddp-server@2.3.2
socket-stream-client@0.3.1
```

**Comment by StorytellerCZ** [comment]
Recently there have been updates to some of these packages. Does the issue still persists?

**Comment by StorytellerCZ** [comment]
I have been trying the reproduce it on the latest Firefox, but without success and the Firefox issue is closed. but #11386 just recently came in so it might still be happening. @mibto can you try to reproduce the issue on Firefox Developer's edition?

**Comment by softwarecreations** [comment]
I can confirm the issue appears to be resolved on Firefox Developer Edition 91.0b3 (64-bit) and Meteor 2.2

**Comment by StorytellerCZ** [comment]
Yes, that is likely. Since it looks like a new version of Firefox fixes the issue I'm closing the issue. Feel free to post here if this re-appears in new versions of Firefox.

**Comment by Divyarajsinh1312** [comment]
I also face the same problem
I make my project in Django
This is my code

```
const chatSocket = new WebSocket("ws://localhost:8000");
	chatSocket.onopen = function (e) {
		console.log("The connection was setup successfully !");
	};
	chatSocket.onclose = function (e) {
		console.log("Something unexpected happened !");
	};
	document.querySelector("#id_message_send_input").focus();
	document.querySelector("#id_message_send_input").onkeyup = function (e) {
		if (e.keyCode == 13) {
		document.querySelector("#id_message_send_button").click();
		}
	};
	document.querySelector("#id_message_send_button").onclick = function (e) {
		var messageInput = document.querySelector(
		"#id_message_send_input"
		).value;
		chatSocket.send(JSON.stringify({ message: messageInput, username : "{{request.user.username}}"}));
	};
	chatSocket.onmessage = function (e) {
		const data = JSON.parse(e.data);
		var div = document.createElement("div");
		div.innerHTML = data.username + " : " + data.message;
		document.querySelector("#id_message_send_input").value = "";
		document.querySelector("#id_chat_item_container").appendChild(div);
	};
```

**Comment by Grubba27** [comment]
hey @Divyarajsinh1312 maybe change new WebSocket("ws://localhost:8000"); to new WebSocket("ws://localhost:8000/socket");

**Comment by lockieluke** [comment]
According to this [stackoverflow post](https://stackoverflow.com/questions/48904150/firefox-can-t-establish-a-connection-to-the-server-at-ws-localhost8080), `const connection = new WebSocket('ws://localhost:8080/');` has to be changed to `const connection = new WebSocket('ws://localhost:8080/', 'echo-protocol');`

**Comment by MBerka** [comment]
> According to this [stackoverflow post](https://stackoverflow.com/questions/48904150/firefox-can-t-establish-a-connection-to-the-server-at-ws-localhost8080), `const connection = new WebSocket('ws://localhost:8080/');` has to be changed to `const connection = new WebSocket('ws://localhost:8080/', 'echo-protocol');`

Unrelated - "echo-protocol" was used by back-end code in the linked question, not in this issue.

**Comment by hatemets** [comment]
Although this might not be relevant for everyone, I managed to get it working by adding a socket `readyState` check in the return statement of the `useEffect` (in React) such as this:
```
useEffect(() => {
        const socket = new WebSocket('ws://localhost:3002');

        return () => {
            if (socket.readyState === 1) {
                socket.close();
            }
        };
    }, []);
```
